integration-tests/src/main/scala/za/co/absa/spline/test/ProducerModelImplicits.scala

```diff
@@ -53,4 +68,34 @@ object ProducerModelImplicits {
     }
   }
 
+  private def findLeaves(operation: Operation)(implicit walker: LineageWalker): (Seq[ReadOperation], Seq[DataOperation]) = {
+
+    @tailrec
+    def findLeavesRec(
+      toVisit: List[Operation],
+      readLeaves: List[ReadOperation],
+      dataLeaves: List[DataOperation]
```

Comment:
Why do you `List` instead of `Seq` here?

Reply:
- List syntax is nice match for this type of algorithm with lot of head tail operations, usable and readable
- List as data structure also works well for what I am trying to do
- Don't see a reason not to

Reply:
What has caught my eye is the `toList` call on a `Seq` in two places.
I understand your intention, and of course I licked am being nitpicking here as in reality those collections in 90-ish % of cases are too small for their type to make any perceivable impact on performance. But since we are talking about it judging from the given method code it doesn't look persuading that you are benefiting from the `List` collection when dealing with the `toVisit` variable, as it in fact might be anything prior converting to a `List`. So every time I see the collection type conversion the question "Why?" pops up in my head. You never know the cost of conversion of a more generic collection type to a more specific, so to me the rule of thumb is that there always have to be really good reasons to do so.

I agree about the rest two variables though.

Reply:
But isn't by the same logic potential cost of head, tail or append on an unknown collection also problematic? The conversion happens just once for each children collection, but the other operations are called for each element in that collection, so multiple times.

Reply:
Well, according to [this](https://docs.scala-lang.org/overviews/collections/performance-characteristics.html) the majority standard immutable collections execute `head` in a constant (or effectively constant) time, while conversion (that most likely requires iteration) will very probably be `O(n)`. The complexity of the whole function seems to be linear as well. To me that means that you would hardly make the overall performance worse if you leave generic `Seq`. And if you do convert, although you won't make it worth either from the `O` perspective, but you may likely increase a constant factor of `N` of `findLeavesRec`'s complexity,. For what it's worth, of course.